pulsar-client/src/main/java/org/apache/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -1243,18 +1243,22 @@ private void closeConsumerTasks() {
         }
         negativeAcksTracker.close();
         stats.getStatTimeout().ifPresent(Timeout::cancel);
-        if (poolMessages) {
-            releasePooledMessagesAndStopAcceptNew();
-        }
+        processIncomingMessagesWhenClose();
     }
 
     /**
-     * If enabled pooled messages, we should release the messages after closing consumer and stop accept the new
+     * when consumer is closed.we should
+     * 1. terminate incomingMessages queue
+     * 2. If enabled pooled messages, we should release the messages after closing consumer and stop accept the new
      * messages.
      */
-    private void releasePooledMessagesAndStopAcceptNew() {
-        incomingMessages.terminate(message -> message.release());
-        clearIncomingMessages();
+    private void processIncomingMessagesWhenClose() {
+        if (poolMessages) {
+            incomingMessages.terminate(message -> message.release());
+            clearIncomingMessages();
+        } else {
+            incomingMessages.terminate(null);
+        }
```

Comment:
btw. it's odd that "clearIncomingMessages" is only called in the case of poolMessage. Since message.release() is a no-op when the message isn't a pooled message, it would be fine to remove the condition for poolMessages.
The implication of not calling "clearIncomingMessages" in all cases is that it won't release memory back to the memory controller for messages that are in the queue. That might also be a problem with the terminate callback.